roles/README.adoc

```diff
@@ -113,6 +113,8 @@ See also <<vars-vs-defaults,Vars vs Defaults>> which mandates "Avoid embedding l
 Since distribution-specific values are kind of "magic values", it applies to them.
 The same logic applies for providers: a role can load a provider-specific variable file, include a provider-specific task file, or both, as needed.
 Consider making paths to templates internal variables if you need different templates for different distributions.
+Do not use roles as programming interfaces, in the case a common role is necessary to contain common code available to other roles within the same collection, it should not contain user-facing plugins.
```

Comment:
I don't grok exactly what you're trying to say.

Reply:
I agree what this is aying isn't very clear.

Reply:
I think we might be better to provide some "do do this" rather than a "dont' do this....

Some things that come to mind:

1) Strive to author loosely coupled hierarchical content, ravioli vs. spaghetti or lasagna (re pasta theory)

2) Place content common to multiple roles in a single role within the collection, "common" is a common name for this role.

3) Limit the consumer's need to understand specific implementation details about a collection, the role names within, and the file names. Presenting the collection as a low-code/no-code "automation application" provides the developer flexibility as the content grows and matures, and limits change the consumer may have to make in a later version.  As an example, #2 here may prove to be a fragile approach and more difficult to maintain long term and requires the consumer have specific details about the roles, files, and structure of the collection

```

- hosts: all
  gather_facts: false
  tasks:
  - name: Perform several actions
    include_role: mycollection.run
    vars:
      actions:
      - name: thing_1
        vars:
          thing_1_var_1: 1
          thing_1_var_2: 2
      - name: thing_2
        vars:
          thing_2_var_1: 1
          thing_2_var_2: 2

- hosts: all
  gather_facts: false
  tasks:
  - name: Do thing_1
    include_role: 
      name: mycollection.thing_1
      tasks_from: thing_1.yaml
      vars:
        thing_1_var_1: 1
        thing_1_var_2: 2
  - name: Do thing_2
    include_role: 
      name: mycollection.thing_2
      tasks_from: thing_2.yaml
      vars:
        thing_2_var_1: 1
        thing_2_var_2: 2
```
4) Provide sane defaults that work for most consumers, minimizing the number of required parameters required for use of the content. Provide configurable opt-in parameters when needed for flexibility.
5) Define the default values for user-configurable parameters in a roles' defaults. This is a familiar place to see and document why certain defaults were chosen and they play into the outcome of the automation.


I think some of these overlap with other areas, but this is what came to mind :)

Reply:
Thanks @cidrblock.  I believe the doc currently covers the points about default, reworked the other suggestions into the doc format.

Reply:
FWIW Point 4 (and to an extent 5) in @cidrblock's response is convention over configuration design that I believe we need to apply a lot more than we have been to Ansible: https://en.wikipedia.org/wiki/Convention_over_configuration.